src/olympia/addons/views.py

```diff
@@ -525,6 +526,96 @@ def get_queryset(self):
             queryset = queryset.transform(Version.transformer_license)
         return queryset
 
+    def create(self, request, *args, **kwargs):
+        addon = self.get_addon_object()
+        has_source = request.data.get('source')
+        if has_source:
+            timer = StopWatch('addons.views.AddonVersionViewSet.create.')
+            timer.start()
+            log.info(
+                'create, source upload received, addon.slug: %s',
+                addon.slug,
+            )
+            timer.log_interval('1.source_received')
+
+        serializer = self.get_serializer(data=request.data)
+        if has_source:
+            log.info(
+                'create, serializer loaded, addon.slug: %s',
+                addon.slug,
+            )
+            timer.log_interval('2.serializer_loaded')
+
+        serializer.is_valid(raise_exception=True)
+        if has_source:
+            log.info(
+                'create, serializer validated, addon.slug: %s',
+                addon.slug,
+            )
+            timer.log_interval('3.serializer_validated')
+
+        self.perform_create(serializer)
+        if has_source:
+            log.info(
+                'create, data saved, addon.slug: %s',
+                addon.slug,
+            )
+            timer.log_interval('4.data_saved')
+
+        headers = self.get_success_headers(serializer.data)
+        return Response(
+            serializer.data, status=status.HTTP_201_CREATED, headers=headers
+        )
+
+    def update(self, request, *args, **kwargs):
```

Comment:
A thought I had while looking at these two chunks of code (that copy DRF source exactly, and are near duplicates in terms of logging - both generally bad things) is whether it would be better to wrap some of the individual functions that both `create` and `update` call instead.

e.g.
``` python
    def get_serializer(self, *args, **kwargs):
        has_source = request.data.get('source')
        if has_source:
            addon = self.get_addon_object()
            self.timer = StopWatch('addons.views.AddonVersionViewSet.update.')
            self.timer.start()
            log.info(
                'update, source upload received, addon.slug: %s, version.id: %s',
                addon.slug,
                instance.id,
            )
            self.timer.log_interval('1.source_received')
        serializer = super().get_serializer(*args, **kwargs)
        if has_source:
            log.info(
                'update, serializer loaded, addon.slug: %s, version.id: %s',
                addon.slug,
                instance.id,
            )
            self.timer.log_interval('2.serializer_loaded')
```
(If differentiating create from update, or logging the version id, was strictly necessary you'd need to tinker with the logging strings to be slightly different if there was an instance or not)

I'm really not sure if that'd actually be any better - the amount of code won't be far off because of the extra conditionals, but it'd all only be defined once and without copying the code in create or update.  Just throwing it out there 🤷‍♂️

Reply:
I had originally through about trying not to override `create` and `update` and instead add timings to functions that they call, as you suggest, but it turned out that there were enough differences, and places where I needed to add timings to `create` and `update` themselves that this just seemed like a better route. I could do what you suggest above, and I could also override `serializer.is_valid` and `serializer.save` to add a timings to those, but I am worried that that will be an equivalent amount of code and will make everything more convoluted, especially if some of the code is in the view and some is in the serializer. 

So, while it's not 100% ideal, I do think what I've done in this patch is probably as good as the alternative. WDYT?